From 41dffd9e16bd15cfb2c2c3dede2d13f46f51be17 Mon Sep 17 00:00:00 2001
From: "Patel, Ankurkumar G" <ankurkumar.g.patel@intel.com>
Date: Mon, 27 Mar 2023 14:52:10 +0530
Subject: [PATCH] USB device enumeration failed with yellow bang

Unknown USB Device (Device Descriptor Request Failed) yellow bang
is observed  in device manager under Universal Serial Bus
controllers.

On AST2600 by default USB multi function pin control(0x440h) is
getting programmed to 00 (USB 1.1 HID controller), Which is
making the USB device configured as unknown under device manager.

Updated to have USB multi function pin control to be programmed
to enable USB2.0 device controller.

Signed-off-by: Patel, Ankurkumar G <ankurkumar.g.patel@intel.com>
---
 arch/arm/mach-aspeed/ast2600/platform.S | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/arch/arm/mach-aspeed/ast2600/platform.S b/arch/arm/mach-aspeed/ast2600/platform.S
index ef272099b4..911b44ab9a 100644
--- a/arch/arm/mach-aspeed/ast2600/platform.S
+++ b/arch/arm/mach-aspeed/ast2600/platform.S
@@ -48,6 +48,7 @@
 #define AST_SCU_DEBUG_CTRL2             (AST_SCU_BASE + 0x0D8)
 #define AST_SCU_HPLL_PARAM		(AST_SCU_BASE + 0x200)
 #define AST_SCU_HPLL_PARAM_EXT		(AST_SCU_BASE + 0x204)
+#define AST_SCU_USB_MULTI_FUNC		(AST_SCU_BASE + 0x440)
 #define AST_SCU_HW_STRAP1		(AST_SCU_BASE + 0x500)
 #define AST_SCU_HW_STRAP1_CLR		(AST_SCU_BASE + 0x504)
 #define AST_SCU_HW_STRAP2		(AST_SCU_BASE + 0x510)
@@ -364,6 +365,11 @@ wait_lock:
 	ldr	r1, =AST_SCU_CA7_PARITY_CHK
 	str	r0, [r1]
 
+	/* select USB2.0 device mode as USB port B */
+	ldr r0, =0x10000000
+	ldr r1, =AST_SCU_USB_MULTI_FUNC
+	str r0, [r1]
+
 #ifdef CONFIG_HW_WATCHDOG
 	/* Enable WDT2 to recover u-boot hang */
 	ldr   r0, =AST_WDT2_RELOAD_VAL
-- 
2.34.1

